x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/CompositeBucketsChangeCollector.java

```diff
@@ -492,20 +670,26 @@ public boolean processSearchResponse(final SearchResponse searchResponse) {
             return true;
         }
 
-        final CompositeAggregation agg = aggregations.get(compositeAggregation.getName());
-
-        Collection<? extends Bucket> buckets = agg.getBuckets();
-        afterKey = agg.afterKey();
+        boolean isDone = true;
 
-        if (buckets.isEmpty()) {
-            return true;
+        // collect changes from composite buckets
+        if (compositeAggregation != null) {
+            final CompositeAggregation agg = aggregations.get(compositeAggregation.getName());
+            Collection<? extends Bucket> buckets = agg.getBuckets();
+            afterKey = agg.afterKey();
+            for (FieldCollector fieldCollector : fieldCollectors.values()) {
+                isDone &= fieldCollector.collectChangesFromCompositeBuckets(buckets);
+            }
+        } else {
+            afterKey = AFTER_KEY_MAGIC;
```

Comment:
If the changes detector is not using a composite agg, the caller shouldn't attempt to iterate on it. Consequently a `null` after key would be acceptable as it would only be called once.

Reply:
This is basically a state problem. I need to know whether the "changes query" has been executed or not, that means this is conceptually a boolean `hasRun`. However, this state is managed by the indexer, not by the change collector. E.g. what if something fails, the change collector doesn't know how to handle failures, but the indexer does. With that magic I let the indexer control this state object instead of having a boolean in the collector.

Having that said, I fully agree that this looks hacky. I will look at it again, suggestions welcome.

Reply:
Could these methods return `IterationResult` instead of a map?

It seems we always need two pieces of info:

- Are we done?
- What is the position?

That might prove a nicer abstraction than having a special bucket position. 

Also, I am not sure in what case the bucket processor (when not utilizing composite aggs), should ever be called twice. It seems weird to me as it should indicate that it is done :/.

Reply:
by moving the `afterKey` to the indexer - where it actually belongs - I was able to remove the bool and the method to get the key. Note that the key can not be persisted right after the response, but only after the real changes got indexed. This is necessary in the failure scenario.

The "special bucket position" is part of making change collection scalable, it's required if more than 65k (10k in older versions) changes have been found.

In case you have 1 field collector that uses min/max and 1 that uses the composite agg(terms), the collector only adds the min/max aggs on the 1st run, but not on following runs in case the terms collector requires it.

I think the challenge is that we combine 2 different ways of change collection in 1 call.